Add command "kubectl config get-contexts"

asalkeld opened the pull request:
``` release-note
* A new command "kubectl config get-contexts" has been added.
```

fixes #25383

bgrant0607 commented on pkg/kubectl/cmd/config/list_context_test.go:
```diff
@@ -0,0 +1,88 @@
+			"shaker-context":     {AuthInfo: "green-user", Cluster: "horse-cluster", Namespace: "chisel-ns"}}}
+	test := listContextTest{
+		startingConfig: tconfig,
+		expectedOut:    "queen-anne-context\nshaker-context\n",
```
Please ensure that the contexts are sorted in a deterministic order.

bgrant0607 replied:
It would also be useful to see the context properties (AuthInfo, Cluster, Namespace), in columnar form.

bgrant0607 commented on pkg/kubectl/cmd/config/list_context.go:
```diff
@@ -0,0 +1,58 @@
+
+	cmd := &cobra.Command{
+		Use:   "list-context",
+		Short: "List the current-contexts",
```
It's all contexts not just the current-context

deads2k commented:
@bgrant0607 This would be another command that is "different" as noted in https://github.com/kubernetes/kubernetes/issues/20605.  Just want to make sure you're ok with another one before we go too far down this path.

bgrant0607 commented on pkg/kubectl/cmd/config/list_context.go:
```diff
@@ -0,0 +1,58 @@
+	options := &ListContextOptions{configAccess: configAccess}
+
+	cmd := &cobra.Command{
+		Use:   "list-context",
```
Please call it `get-context`. `list` is not an ordinary kubectl command, but `get` is.

`kubectl get-context` should print all contexts.

`kubectl get-context <context-name>` should print just the selected context.

`kubectl get-context -o name` should print the names of all contexts.

bgrant0607 commented:
@deads2k I'm ok with it, though we should try to make it as similar to other commands as we can, hence my comments about get vs. list, columnar output, etc.

At some point, we'll need to redo the whole command group.

asalkeld commented:
Thanks for the review @bgrant0607 and @deads2k, I'll re-work this to better fit in.

asalkeld commented:
Please let me know if I should squash these commits, there is a lot of renaming going on.

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
2016

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+
+	get_context_example = `
+  $ kubectl config get-context
+    NAME                 CLUSTER         AUTHINFO     NAMESPACE
```
@kargakis remind me.  We don't usually demonstrate output in the examples, right?

kargakis replied:
We don't.

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+	get_context_long = `Displays one or many contexts from the kubeconfig file.`
+
+	get_context_example = `
+  $ kubectl config get-context
```
Add a `# list all the contexts in your kubeconfig file` style explanation for your examples.

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+    federal-context      horse-cluster   green-user   chisel-ns
+    queen-anne-context   big-cluster     blue-user    saw-ns
+
+  $ kubectl config get-context -o name
```
`-o name` for other commands excludes headers

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+    federal-context      horse-cluster   green-user   chisel-ns
+    queen-anne-context   big-cluster     blue-user    saw-ns
+
+  $ kubectl config get-context --no-headers
```
`--no-headers` is standard.  probably don't need an example.  However, `get-context <name>` would be good to show.

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+
+	cmd := &cobra.Command{
+		Use:     "get-context [(-o|--output=)detail|name)]",
+		Short:   "Get one or many contexts",
```
Display instead of get I think.  See `kubectl get` help

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+			cmdutil.CheckErr(options.RunGetContext(cmd, out))
+		},
+	}
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: detail|name.")
```
Add a TODO to collapse onto `AddPrinterFlags`.  This was ok'ed as a one-off, but consistency would still be good.

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+		},
+	}
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: detail|name.")
+	cmd.Flags().Bool("no-headers", false, "Don't print column headers.")
```
Keep this flag description consistent

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+func (o *GetContextOptions) Complete(cmd *cobra.Command, argsIn []string) error {
+	if len(argsIn) == 1 {
```
You should support enumerating more than one context name.

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+		return nil
+	}
+
+	for name := range config.Contexts {
```
Do this check in a run.  You should behave like `kubectl get` when you get invalid names.  You note the failure and return non-zero, but you show all the information you can.

deads2k commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// RunGetContext implements all the necessary functionality for context retrieval.
+func (o GetContextOptions) RunGetContext(cmd *cobra.Command, output io.Writer) error {
```
avoid passing the cmd and the output here.  Those can be values on your options struct.

deads2k replied:
> avoid passing the cmd and the output here. Those can be values on your options struct.

We've done this openshift by passing a printer function that we build in `Complete`.  It keeps the code down in run cleaner.

janetkuo commented:
ok to test

janetkuo commented on pkg/kubectl/cmd/config/get_context.go:
```diff
@@ -0,0 +1,163 @@
+		},
+	}
+	// TODO when the resource printer can print non-runtime objects use AddPrinterFlags here.
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml|wide|name|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://releases.k8s.io/HEAD/docs/user-guide/jsonpath.md].")
```
Suggest creating a function for `output` flag, and call that function here and in `AddPrinterFlags`

janetkuo replied:
and for `no-headers` too (they can be together)

janetkuo commented:
Suggest naming it `get-contexts` (plural)

janetkuo commented:
PR title and release note need to be updated too

janetkuo commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -45,6 +45,16 @@ func AddOutputFlagsForMutation(cmd *cobra.Command) {
 	cmd.Flags().StringP("output", "o", "", "Output mode. Use \"-o name\" for shorter output (resource/name).")
 }
 
+// AddOutputFlagsForMutation adds output related flags to a command.
```
nit: should be `AddOutputFlags`

deads2k commented on pkg/kubectl/cmd/config/get_contexts.go:
```diff
@@ -0,0 +1,161 @@
+		Example: get_contexts_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			outputFormat := cmdutil.GetFlagString(cmd, "output")
+			outputTypes := map[string]bool{"": true, "json": false, "yaml": false, "wide": false, "name": true, "go-template": false, "go-template-file": false, "jsonpath": false, "jsonpath-file": false}
```
`pkg/util/sets.String`

asalkeld replied:
I was asked to have the standard --output help string. But I can't support all the options as resource_printer needs a runtime object. So I wanted to differentiate between when a user gives an option that is in the help string (seemingly legal) and total garbage. In the case of a legal-but-not-supported case I am doing what "kubectl config view" does (reset to the default, and print a message).

deads2k commented on pkg/kubectl/cmd/config/get_contexts.go:
```diff
@@ -0,0 +1,161 @@
+				cmdutil.CheckErr(fmt.Errorf("output must be one of '' or 'name': %v.", outputFormat))
+			}
+			if !supported {
+				fmt.Printf("--output %v is not available in kubectl config get-contexts; reseting to default output format\n", outputFormat)
```
pass in an `outErr io.Writer` and print to that.

deads2k commented:
lgtm

I think its worth a followup issue to decide if the behavior should mirror `git branch` where it highlights the current branch with a `*`.

asalkeld commented:
Rebased and @deads2k it now includes "git branch" style *

```
$ kubectl config get-contexts
CURRENT   NAME      CLUSTER   AUTHINFO   NAMESPACE
          fluffy    local     bear       
*         local     local                
$ kubectl config get-contexts -o name
  fluffy
* local
```

deads2k commented:
> Rebased and @deads2k it now includes "git branch" style *

Sorry, can't do that with `-o name`.  That flag is intended for scripts and a `*` in the output could cause them trouble.

asalkeld commented:
ok, i'll remove it out of -o name.

deads2k commented:
CI says

FAILED   ./hack/../hack/verify-generated-docs.sh	148s

asalkeld commented:
rebased, and fixed the generated docs failure..

Commit:
Add command "kubectl config get-contexts"

Note: Context is not a runtime object (doesn't have Kind and Version) so
      we can't use the resource_printer
